### PR TITLE
Plot grain area distribution in nm

### PR DESCRIPTION
**Pull request by dtorrejo**

### CODE REVIEW

File: methods/main_methods.py [hunk 1]
```diff
@@ -4,19 +4,17 @@
...
 from sklearn.mixture import GaussianMixture
 
 from methods import config
 
 PROPS = config.data_properties
 HEIGHT_INDEX = config.height_index
+SAMPLE_SIZE = {'backgrounded': 1, '2componentfilms': 0.5, 'nanowires': 5} # nano-meters
 
-MAX_PIXEL = 255
 LABEL_THRESH = 1000 # each label must have more than this number of pixels
```
Comment (dtorrejo) [inline methods/main_methods.py]: @wesleyktatum Parameter that you need to lower to plot smaller grains

File: methods/main_methods.py [hunk 2]
```diff
@@ -302,6 +300,7 @@ def get_unique_labels(labels):
     if 0 in unique_labels: # skips outliers AND borders in watershed segmentation
         unique_labels.remove(0)
 
+    unique_labels = sorted(unique_labels, key=lambda k: np.sum(labels==k))
```
Comment (dtorrejo) [inline methods/main_methods.py]: Sorts labels per area and uses same ordering to select colors in plots.

File: methods/main_methods.py [hunk 3]
```diff
@@ -4,19 +4,17 @@
...
 from scipy import signal
-from matplotlib import pyplot
+from matplotlib import pyplot, colors, cm
 from sklearn.mixture import GaussianMixture
 
 from methods import config
 
 PROPS = config.data_properties
 HEIGHT_INDEX = config.height_index
+SAMPLE_SIZE = {'backgrounded': 1, '2componentfilms': 0.5, 'nanowires': 5} # nano-meters
```
Comment (dtorrejo) [inline methods/main_methods.py]: @wesleyktatum Let me know if these values are correct.

File: ssts/segmentation/persistence_watershed_algorithm.py [hunk 4]
```diff
@@ -141,8 +141,11 @@ def merge_tree(maxima, edges):
             if values[vc] < values[uc]:  # if later label has lower max
                 uc, vc = vc, uc
 
-            # source, target, persistence (size of lower label hill)
-            pairs.append((uc, vc, values[uc] - val))
-            components[uc] = components[vc]  # lower label maps to higher label
+            pairs.append((uc, vc, values[uc] - val)) # source, target, persistence (size of lower label hill)
+
+            # NOTE without this constraint, artifacts appear in the output plot
+            if (values[uc] - val) < pers_thresh:
```
Comment (dtorrejo) [inline ssts/segmentation/persistence_watershed_algorithm.py:147]: @ponl This is where the artifacts were coming from. We should talk about this because I don't think this is optimal

File: ssts/segmentation/example.py [hunk 5]
```diff
@@ -35,6 +35,10 @@ def main(data_path=None, example_number=None):
         # Plot classification distributions
         mm.show_classification_distributions(labels, data)
 
+        # Overlay distributions on pixels
+        probs = seg.get_probabilities(data)
+        mm.show_overlaid_distribution(probs, data)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:42]: Shows distributions on top of pixel data.

File: methods/main_methods.py [hunk 6]
```diff
@@ -18,24 +18,31 @@
...
+def show_property_distributions(data, outliers=None):
     """ Plots the pdfs of the data properties
     Args:
         data (np array): data
         outliers (np array): outliers
     """
-    fig = pyplot.figure(figsize=(15, 10), dpi=80, facecolor='w', edgecolor='k')
-
     h, w, c = data.shape
+    num_plots = c
```
Comment (dtorrejo) [inline methods/main_methods.py:30]: automates the set up for plotting (number of columns and rows)

File: ssts/segmentation/example.py [hunk 7]
```diff
@@ -89,6 +89,9 @@ def main(data_path=None, example_number=None):
         pca_components = seg.pca.transform(data.reshape(n, c))  # NOTE PCA was trained while fitting pre_seg
         pca_components = pca_components.reshape(h, w, num_pca_components)  # shape (512, 512, num_pca_components)
 
+        # Plot classification distributions of PCA components
+        mm.show_classification_distributions(pre_labels, pca_components, title_flag=False)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:94]: Show classification distributions of PCA components. You can use this to see what components and properties dominate the classification

File: ssts/segmentation/example.py [hunk 8]
```diff
@@ -188,6 +188,36 @@ def main(data_path=None, example_number=None):
         # Plot classification distributions with masks
         mm.show_distributions_together(labels, data)  # all classes plotted together
 
+    # NOTE Wes workflow
+    elif example_number == "7":
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:196]: I added a new example to match your workflow @wesleyktatum

File: ssts/segmentation/example.py [hunk 9]
```diff
@@ -188,6 +188,36 @@ def main(data_path=None, example_number=None):
...
+
+        no_outliers_data = np.copy(data)
+        no_outliers_data[outliers==1] = 0
+
+        # Run GMM segmentation
+        seg1 = seg_gmm.SegmenterGMM(n_components=2, embedding_dim=5)
+        seg1_labels = seg1.fit_transform(data, outliers)
+
+        # Remove height property (optional)
+        #no_height_data = np.delete(no_outliers_data, 4, axis=2)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:217]: @wesleyktatum You can exclude height property if needed

File: ssts/segmentation/example.py [hunk 10]
```diff
@@ -188,6 +188,36 @@ def main(data_path=None, example_number=None):
...
+        #seg1_labels = seg1.fit_transform(no_height_data, outliers)
+
+        # Plot classification
+        mm.show_classification(seg1_labels, no_outliers_data)
+
+        # Plot classification distributions
+        mm.show_distributions_together(seg1_labels, no_outliers_data)
+
+        # Create unique masks per grain
+        seg2_labels = seg1.get_grains(seg1_labels)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:227]: @wesleyktatum I exchanged the persistence watershed method for the nearest neighbor method as mentioned in our meetings. I'd recommend using this in your experiments.

File: methods/main_methods.py [hunk 11]
```diff
@@ -100,26 +99,71 @@ def show_outliers(data, outliers, height_index=HEIGHT_INDEX):
     pyplot.show()
 
 ## Frequency removal
-def apply_frequency_removal(data):
+def apply_frequency_removal(data, compression_percent=0.25):
```
Comment (dtorrejo) [inline methods/main_methods.py]: Applies low frequency removal from data. Compression_percent (between 0 and 1) dictates how much compression needs to be done. @wesleyktatum

File: ssts/segmentation/example.py [hunk 12]
```diff
@@ -206,6 +187,41 @@ def main(data_path=None, example_number=None):
         # Plot classification distributions with masks
         mm.show_distributions_together(labels, data)  # all classes plotted together
 
+    # NOTE Wes workflow
+    elif example_number == "7":
+
+        # Apply frequency removal
+        data = mm.apply_frequency_removal(data)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:199]: Low frequency removal step

File: ssts/segmentation/example.py [hunk 13]
```diff
@@ -80,14 +88,20 @@ def main(data_path=None, example_number=None):
         pca_components = seg.pca.transform(data.reshape(n, c))  # NOTE PCA was trained while fitting pre_seg
         pca_components = pca_components.reshape(h, w, num_pca_components)  # shape (512, 512, num_pca_components)
 
+        # Plot classification distributions of PCA components
+        mm.show_classification_distributions(pre_labels, pca_components, title_flag=False)
+
+        # Show correlation after classification of PCA components
+        mm.show_classification_correlation(pre_labels, pca_components, title_flag=False)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:97]: @wesleyktatum Scatter plot / correlation of PCA components

File: ssts/segmentation/example.py [hunk 14]
```diff
@@ -206,10 +190,49 @@ def main(data_path=None, example_number=None):
...
+
+        # Show a-priori distributions
+        mm.show_property_distributions(data, outliers)
+
+        # Run GMM segmentation
+        seg1 = seg_gmm.SegmenterGMM(n_components=2)
+        seg1_labels = seg1.fit_transform(data, outliers)
+
+        # Show correlation after classification
+        mm.show_classification_correlation(seg1_labels, no_outliers_data)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:214]: @wesleyktatum Scatter plot / correlation of properties after classification

File: methods/main_methods.py [hunk 15]
```diff
@@ -97,48 +57,118 @@ def extract_outliers(data, height_index=HEIGHT_INDEX, threshold=2.5):
...
+    m = pyplot.imshow(no_outliers_data[:, :, height_index], aspect="auto")
+    pyplot.title("Height")
     pyplot.colorbar(m, fraction=0.046, pad=0.04)
 
     pyplot.tight_layout()
     pyplot.show()
 
+
+## Frequency removal
+def apply_frequency_removal(data, compression_percent=95):
```
Comment (dtorrejo) [inline methods/main_methods.py:123]: Compression is now between 0 and 100

File: methods/main_methods.py [hunk 16]
```diff
@@ -97,48 +57,118 @@ def extract_outliers(data, height_index=HEIGHT_INDEX, threshold=2.5):
...
+            f_prop_shift (np.array): frequencies
+            h (int): height of data array
+            w (int): width of data array
+            compression_percent (float): percentage of compression
+        Returns:
+            f_prop_shift (np.array): high frequencies
+        """
+        mags = np.abs(f_prop_shift)
+        thresh = np.percentile(mags, compression_percent)
+        cond = np.abs(f_prop_shift) < thresh
```
Comment (dtorrejo) [inline methods/main_methods.py:144]: Improved high-filter pass using the magnitude of the data array

File: methods/main_methods.py [hunk 17]
```diff
@@ -409,60 +432,183 @@ def show_distributions_together(labels, data):
...
+    for i in range(n):
+        ax = pyplot.subplot(num_rows, num_cols, cnt)
+        cnt += 1
+        m = ax.imshow(probs[:, :, i], vmin=0, vmax=1)
+        pyplot.colorbar(m, fraction=0.046, pad=0.04)
+
+    pyplot.show()
+
+
+def show_classification_correlation(labels, data, title_flag=True):
```
Comment (dtorrejo) [inline methods/main_methods.py]: Method to plot scatter plots / correlation after classification

File: ssts/segmentation/segmentation_gmm.py [hunk 18]
```diff
@@ -148,6 +156,48 @@ def fit_transform(self, data, outliers=None):
...
+
+        if self.embedding_dim is not None:
+            data = self.pca.transform(data)
+
+        vector_probs = self.gmm.predict_proba(data)
+        probs = vector_probs.reshape(h, w, self.n_components)
+        return probs
+
+    @staticmethod
+    def add_nonlinear_features(data):
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_gmm.py:203]: Adds squared, cubed, absolute value, and reciprocal representations of data as features for classification

File: ssts/segmentation/segmentation_gmm.py [hunk 19]
```diff
@@ -12,7 +12,7 @@
 
 
 class SegmenterGMM(object):
-    def __init__(self, n_components=3, normalize=True, embedding_dim=None, padding=0, zscale=False):
+    def __init__(self, n_components=3, normalize=True, embedding_dim=None, padding=0, zscale=False, nonlinear=False):
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_gmm.py:15]: @wesleyktatum Need to set nonlinear flag to True to add nonlinearities

File: ssts/segmentation/example.py [hunk 20]
```diff
@@ -206,10 +190,49 @@ def main(data_path=None, example_number=None):
...
+        outliers = mm.extract_outliers(data)
+        no_outliers_data = np.copy(data)
+        no_outliers_data[outliers==1] = 0
+        mm.show_outliers(data, outliers)
+
+        # Show a-priori distributions
+        mm.show_property_distributions(data, outliers)
+
+        # Run GMM segmentation
+        seg1 = seg_gmm.SegmenterGMM(n_components=2, nonlinear=True)
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:210]: @wesleyktatum Nonlinear flag set to True

File: ssts/segmentation/example.py (hunk 12 above)
Comment (dtorrejo) [reply ssts/segmentation/example.py:199]: This step has been improved to filter according to frequency magnitude as opposed to within some radius

File: ssts/segmentation/example.py (hunk 20 above)
Comment (dtorrejo) [reply ssts/segmentation/example.py:210]: Note that the pca flag can be set to true and dimensionality reduction will be applied to the nonlinear features.

File: ssts/segmentation/example.py [hunk 21]
```diff
@@ -15,9 +15,10 @@ def main(data_path=None, example_number=None):
         args = p.parse_args()
 
     example_number = args.example_number
-    data_dir = args.data_path
-    data_type = data_dir.split("/")[-3]
-    data = np.load(data_dir)
+    data_file = args.data_path
+    data_type = data_file.split("/")[-3]
+    data_dir = "/".join(data_file.split("/")[:-1])
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:22]: extracts data directory from specified data file

File: ssts/segmentation/example.py [hunk 22]
```diff
@@ -228,6 +229,10 @@ def main(data_path=None, example_number=None):
         # Plot grain classification
         mm.show_classification(seg2_labels, no_outliers_data)
 
+    # NOTE A-priori correlation analysis
+    elif example_number == "8": # NOTE this looks better the more data files in the dir
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:233]: Example shows how to get correlation information across data files (inside the same directory)

File: ssts/segmentation/segmentation_gmm.py [hunk 23]
```diff
@@ -111,22 +111,8 @@ def transform(self, data, outliers=None):
...
-            data = SegmenterGMM.get_windows(data, padding=self.padding)
-        else:
-            data = data.reshape(n, c)
-
-        if self.normalize:
-            data = self.sst.transform(data) if self.zscale else SegmenterGMM.normalize_by_max(data)
-
-        if self.embedding_dim is not None:
-            data = self.pca.transform(data)
+        data = self.get_pca_components(data)
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_gmm.py:115]: @wesleyktatum I've created a method to extract pca_components seamlessly after fitting model

**Comment by dtorrejo** [comment]
@wesleyktatum I have added the Jupyter notebook with all the current capabilities

File: methods/main_methods.py [hunk 24]
```diff
@@ -576,7 +576,8 @@ def show_classification_correlation(labels, data, title_flag=True, sample_flag=T
 
                 if sample_flag:
                     indices = np.arange(data_i.shape[0])
-                    sampling = np.random.choice(indices, size=1000, replace=False)
+                    sample_size = min(1000, len(indices))
```
Comment (dtorrejo) [inline methods/main_methods.py:597]: @wesleyktatum This should deal with lower data samples.

File: methods/main_methods.py [hunk 25]
```diff
@@ -91,16 +91,34 @@ def show_outliers(data, outliers, height_index=HEIGHT_INDEX):
...
-    m = pyplot.imshow(no_outliers_data[:, :, height_index], aspect="auto")
+    m = pyplot.imshow(height_data, aspect="auto")
     pyplot.title("Height")
     pyplot.colorbar(m, fraction=0.046, pad=0.04)
 
     pyplot.tight_layout()
     pyplot.show()
 
 
+def smooth_outliers_from_data(data, outliers):
```
Comment (dtorrejo) [inline methods/main_methods.py:105]: @wesleyktatum Maps outliers of each properties to their mean value.

File: ssts/segmentation/segmentation_gmm.py [hunk 26]
```diff
@@ -222,7 +222,7 @@ def get_windows(data, padding=3, flatten=True):
         n = 2 * padding + 1
         wins = np.zeros((h * w, n ** 2, c))
         for d in range(c):
-            X = np.pad(data[:, :, d], padding, "reflect")
+            X = np.pad(data[:, :, d], padding, "symmetric")
```
Comment (dtorrejo) [inline ssts/segmentation/segmentation_gmm.py:225]: @wesleyktatum  Before we were using "constant" or "reflective" which was adding constant values to the boundaries. Now I changed that to reflect the interior value on the exterior. This removes the frame artifact you were observing.

File: ssts/segmentation/example.py [hunk 27]
```diff
@@ -5,6 +5,8 @@
 import ssts.segmentation.segmentation_watershed as seg_water
 from methods import main_methods as mm  # NOTE SSTS directory must be in PYTHONPATH
 
+Z_SCORE_THRESH = 2.5
```
Comment (dtorrejo) [inline ssts/segmentation/example.py:8]: @wesleyktatum This is how you change the value of the z score threshold used for outlier extraction.
